k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/FolderSelectedStateCommand.java

```diff
@@ -0,0 +1,172 @@
+package com.fsck.k9.mail.store.imap.selectedstate.command;
+
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.store.imap.ImapConnection;
+import com.fsck.k9.mail.store.imap.ImapFolder;
+import com.fsck.k9.mail.store.imap.ImapResponse;
+import com.fsck.k9.mail.store.imap.ImapUtility;
+import com.fsck.k9.mail.store.imap.selectedstate.response.SelectedStateResponse;
+
+
+abstract class FolderSelectedStateCommand {
+
+    /* The below limits are 20 octets less than the recommended limits, in order to compensate for
+    the length of the command tag, the space after the tag and the CRLF at the end of the command
+    (these are not taken into account when calculating the length of the command). For more
+    information, refer to section 4 of RFC 7162.
+
+    The length limit for servers supporting the CONDSTORE extension is large in order to support
+    the QRESYNC parameter to the SELECT/EXAMINE commands, which accept a list of known message
+    sequence numbers as well as their corresponding UIDs.
+     */
+    private static final int LENGTH_LIMIT_WITHOUT_CONDSTORE = 980;
+    private static final int LENGTH_LIMIT_WITH_CONDSTORE = 8172;
+
+    private Set<Long> idSet;
+    private List<ContiguousIdGroup> idGroups;
+
+    FolderSelectedStateCommand(Set<Long> idSet) {
+        this.idSet = new TreeSet<>(idSet);
+        this.idGroups = new ArrayList<>(0);
+    }
+
+    abstract String createCommandString();
+
+    String createCombinedIdString() {
+        if (idSet.isEmpty() && idGroups.isEmpty()) {
+            return "";
+        }
+
+        StringBuilder builder = new StringBuilder();
+
+        builder.append(ImapUtility.join(",", idSet));
+        if (!idSet.isEmpty() && !idGroups.isEmpty()) {
+            builder.append(",");
+        }
+        builder.append(ImapUtility.join(",", idGroups));
+
+        builder.append(" ");
+        return builder.toString();
+    }
+
+    public SelectedStateResponse execute(ImapConnection connection, ImapFolder folder) throws MessagingException {
+        return null;
+    }
+
+    List<List<ImapResponse>> executeInternal(ImapConnection connection, ImapFolder folder)
```

Comment:
I don't quite like that the responsibility for executing the actual call on the `ImapConnection` lies in this object. I'd rather this be part of an `ImapConnection.execute` method.

As a suggestion:
```
abstract class <R extends FolderSelectedResponse> FolderStateSelectedCommand<R> {
  abstract R parseResponse(List<ImapResponse> unparsedResponses);
}

public class ImapConnection {
  public <R extends FolderSelectedResponse> R execute(FolderSelectedStateCommand<R> command) {
    // roughly what executeInternal does
    // obtains the list of commands from command.getCommands
    // parses the response from command.parseResponse to get the result
  }
}
```

That way, the command classes only provide the commands as strings, and a way to parse the responses into specialized objects. Thoughts?

Reply:
@Valodim I'm not sure if the parsing logic should be in the Command or the Response classes, both seem fine to me. I'm leaning slightly towards the response classes though.